### PR TITLE
Homework 4 - SECOND ENTITY AND MANY-TO-MANY ENTITY RELATIONSHIPS

### PR DESCRIPTION
**Pull request by helenakolodko**
# SECOND ENTITY AND MANY-TO-MANY ENTITY RELATIONSHIPS

## TASK 4.1

Add **Group** entity to already existing **REST** service with **CRUD** operations.

The **Group** entity should have the following properties (you can use **UUID** as Group **id**):

```typescript
type Permission = 'READ' | 'WRITE' | 'DELETE' | 'SHARE' | 'UPLOAD_FILES';

type GroupInterface = {
    id: string;
    name: string;
    permissions: Array<Permission>;
};
```

The service should provide the following **CRUD** operations for **Group**:
- get group by **id**;
- get all groups;
- create and update a group;
- remove group (**hard delete** – group data is fully removed from the DB).
Storing of groups data should be done in **PostgreSQL** in **Groups** table.
The service should follow the principles of 3-layer architecture.

## TASK 4.2

Link User records in one table with Group records in another table.
- Add a **UserGroup** table (*“many-to-many”* relationship) which will store the data describing
which users are assigned to which group.
- If any record gets removed from the DB, then all linked records should be removed from
**UserGroup** as well.

## TASK 4.3

Add `addUsersToGroup(groupId, userIds)` method which will allow adding users to a certain group. Use transactions to save records in DB.

# EVALUATION CRITERIA

2. Group entity is added, operations are implemented but there are some code structure/quality issues.
3. Task 4.1 is fulfilled to the full extent.
4. Task 4.2 is fulfilled to the full extent.
5. Task 4.3 is fulfilled to the full extent; transactions are used for saving records.

To check the solution:

1. fill `.env` file with your DB credentials
2. run SQL script `api\src\scripts\users.sql` to create required tables
3. run `npm run task`

### CODE REVIEW

File: api/src/entities/user/user.service.ts [hunk 1]
```diff
@@ -49,14 +50,14 @@ export class UserService {
     }
 
     softDelete = async (user: UserInterface) => {
-        const deleted = await User.update({ isDeleted: true }, {
-            where: {
-                id: user.id
-            }
+        Connection.transaction(async transaction => {
+            await User.update({ isDeleted: true }, { where: { id: user.id }, transaction });
+            await UserGroup.destroy({ where: { userId: user.id }, transaction });
```
Comment (RockMinsk) [inline api/src/entities/user/user.service.ts:55]: According to task 4.2 (point 2):
> If any record gets removed from the DB, then all linked records should be removed from UserGroup as well.

So why do you use 'destroy' method and delete from database dependencies in  this softDelete method that doesn't delete anything from database?

**Review by RockMinsk — COMMENTED** [review]
Please take a look at notes and fix issues

File: api/src/entities/user/user.service.ts (hunk 1 above)
Comment (helenakolodko) [reply api/src/entities/user/user.service.ts]: I just read this task incorrectly. Removed this from soft delete method

File: api/src/validation/groupExistsValidation.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { Request, Response, NextFunction } from 'express';
+import {GroupService} from '../entities/group/group.service';
+
+const groupExistsValidation = (groupService: GroupService) => {
+    return async (req: Request, res: Response, next: NextFunction) => {
+        const logins = await groupService.allGroupNames();
+
+        if (logins.includes(req.body.login)) {
+            res.status(400)
+                .json(`Group with name '${req.body.login}' already exists.`);
```
[the remark this answers is not in the document]
Comment (helenakolodko) [reply api/src/validation/groupExistsValidation.ts]: Fixed the typo

File: api/src/entities/group/group.service.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                        groupId: groupId,
+                    },
+                    {
+                        transaction,
+                    }
+                );
+            });
+            await Promise.all(added);
+        });
+    }
```
[the remark this answers is not in the document]
Comment (helenakolodko) [reply api/src/entities/group/group.service.ts:42]: I used different body format. Fixed this and added validation

File: api/src/validation/groupSchema.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const Joi = require('@hapi/joi');
+import { Permission } from '../entities/group/group.interface';
+
+const userSchema = Joi
+    .object({
+        id: Joi.string()
+            .guid({ version : 'uuidv4' })
+            .required(),
```
[the remark this answers is not in the document]
Comment (helenakolodko) [reply api/src/validation/groupSchema.ts]: Added different validation for create and update operations. Fixed issue for regeneration of provided UUID

File: api/src/db/postgresConnection.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+export const Connection = new Sequelize(
+    process.env.DB_NAME || 'users',
+    process.env.DB_USER || 'admin',
+    process.env.DB_PASSWORD || 'password',
+    {
+      host: process.env.DB_HOST || 'localhost',
+      dialect: 'postgres'
+    }
+);
```
[the remark this answers is not in the document]
Comment (helenakolodko) [reply api/src/db/postgresConnection.ts]: Good point. Fixed it

File: api/src/db/models/group.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        },
+        permissions: {
+            allowNull: false,
+            type: DataTypes.ARRAY(DataTypes.STRING)
+        }
+    },
+    {
+        sequelize: Connection,
+        timestamps: false,
+        schema: 'public',
```
[the remark this answers is not in the document]
Comment (helenakolodko) [reply api/src/db/models/group.ts]: Fixed

File: api/src/app.ts [hunk 7]
```diff
@@ -1,17 +1,24 @@
...
+import groupRouter from './entities/group/group.routes';
+import dotenv from 'dotenv';
+import { Connection } from './db/postgresConnection';
+
+dotenv.config();
 
 const app: express.Application = express();
 
-app.listen(3000, () => {
+//Connection.sync();
```
[the remark this answers is not in the document]
Comment (helenakolodko) [reply api/src/app.ts]: Just commented it for some testing purposes and forgot about it. Fixed

File: api/Readme.md [hunk 8]
```diff
@@ -58,4 +58,42 @@ The service should adhere to 3-layer architecture principles (https://softwareon
...
+
+Link User records in one table with Group records in another table.
+- Add a **UserGroup** table (*“many-to-many”* relationship) which will store the data describing
+which users are assigned to which group.
+- If any record gets removed from the DB, then all linked records should be removed from
+**UserGroup** as well.
+
+## TASK 4.3
+
+Add `addUsersToGroup(groupId, userIds)` method which will allow adding users to a certain group. Use transactions to save records in DB.
```
[the remark this answers is not in the document]
Comment (helenakolodko) [reply api/Readme.md:99]: Added to README

File: api/.env [hunk 9]
```diff
@@ -0,0 +1,5 @@
+DB_HOST=
+DB_USER=
+DB_PASSWORD=
+DB_NAME=
+APP_PORT=APP_PORT=3000
```
[the remark this answers is not in the document]
Comment (helenakolodko) [reply api/.env]: Fixed

**Pull request merged**